platforms/macos/src/subclass.rs

```diff
@@ -0,0 +1,177 @@
+use parking_lot::Mutex;
+use std::{collections::HashMap, ffi::c_void, ops::Deref};
+
+use crate::{appkit::NSView, Adapter};
+
+static SUBCLASSES: Lazy<Mutex<HashMap<&'static Class, &'static Class>>> =
+    Lazy::new(|| Mutex::new(HashMap::new()));
+
+struct Instance {
+    view: Id<NSView, Shared>,
+    adapter: Adapter,
+    prev_class: Option<&'static Class>,
+}
+
+#[derive(PartialEq, Eq, Hash)]
+struct ViewKey(*const NSView);
+unsafe impl Send for ViewKey {}
+unsafe impl Sync for ViewKey {}
+
+struct InstancePtr(*const Instance);
+unsafe impl Send for InstancePtr {}
+unsafe impl Sync for InstancePtr {}
+
+static INSTANCES: Lazy<Mutex<HashMap<ViewKey, InstancePtr>>> =
+    Lazy::new(|| Mutex::new(HashMap::new()));
+
+// Some view classes, like the one in winit 0.27, assume that they are the
+// lowest subclass, and call [self superclass] to get their superclass.
+// Give them the answer they need.
+unsafe extern "C" fn get_superclass(this: &NSView, _cmd: Sel) -> Option<&Class> {
+    let key = ViewKey(this as *const _);
+    let instances = INSTANCES.lock();
+    let instance = instances.get(&key).unwrap();
+    (*instance.0).prev_class.as_ref().unwrap().superclass()
+}
+
+unsafe extern "C" fn children(this: &NSView, _cmd: Sel) -> *mut NSArray<NSObject> {
+    let key = ViewKey(this as *const _);
+    let instances = INSTANCES.lock();
+    let instance = instances.get(&key).unwrap();
+    (*instance.0).adapter.view_children()
+}
+
+unsafe extern "C" fn focus(this: &NSView, _cmd: Sel) -> *mut NSObject {
+    let key = ViewKey(this as *const _);
+    let instances = INSTANCES.lock();
+    let instance = instances.get(&key).unwrap();
+    (*instance.0).adapter.focus()
+}
+
+unsafe extern "C" fn hit_test(this: &NSView, _cmd: Sel, point: NSPoint) -> *mut NSObject {
+    let key = ViewKey(this as *const _);
+    let instances = INSTANCES.lock();
+    let instance = instances.get(&key).unwrap();
+    (*instance.0).adapter.hit_test(point)
+}
+
+impl Instance {
+    fn new(view: Id<NSView, Shared>, adapter: Adapter) -> Box<Self> {
+        Box::new(Self {
+            view,
+            adapter,
+            prev_class: None,
+        })
+    }
+
+    fn install(&mut self) {
+        let view_ptr = Id::as_ptr(&self.view);
+        let key = ViewKey(view_ptr);
+        INSTANCES.lock().insert(key, InstancePtr(self as *const _));
+        // Cast to a pointer and back to force the lifetime to 'static
+        // SAFETY: We know the class will live as long as the instance,
+        // and we own a reference to the instance.
+        let superclass = unsafe { &*(self.view.class() as *const Class) };
+        self.prev_class = Some(superclass);
+        let mut subclasses = SUBCLASSES.lock();
+        let entry = subclasses.entry(superclass);
+        let subclass = entry.or_insert_with(|| {
+            let name = format!("AccessKitSubclassOf{}", superclass.name());
+            let mut builder = ClassBuilder::new(&name, superclass).unwrap();
```

Comment:
Re manual class declaration:

While it'd of course be cleanest if we could use `declare_class!`, I think it's okay if you have to drop down into the manual stuff sometimes. (I'll probably have a bigger document at some point describing this philosophy of "layered safety").